src/internal/DotStarEsp32DmaSpiMethod.h

```diff
@@ -0,0 +1,320 @@
+
+        // must have a 4 byte aligned buffer for i2s
+        uint32_t alignment = _spiBufferSize % 4;
+        if (alignment)
+        {
+            _spiBufferSize += 4 - alignment;
+        }
+
+        _data = static_cast<uint8_t*>(malloc(_spiBufferSize));
+        _dmadata = static_cast<uint8_t*>(heap_caps_malloc(_spiBufferSize, MALLOC_CAP_DMA));
+
+        // data cleared later in NeoPixelBus::Begin()
+    }
+
+    // Support constructor specifying pins by ignoring pins
+    DotStarEsp32DmaSpiMethod(uint8_t, uint8_t, uint16_t pixelCount, size_t elementSize, size_t settingsSize) :
+        DotStarEsp32DmaSpiMethod(pixelCount, elementSize, settingsSize)
+    {
+    }
+
+    ~DotStarEsp32DmaSpiMethod()
+    {
+        if(_spiHandle) {
+            DeInitSpiDevice();
+            esp_err_t ret=spi_bus_free(T_SPIBUS::spiHostDevice);
+            ESP_ERROR_CHECK(ret);            
+        }
+        free(_data);
+        free(_dmadata);
+        _spiHandle = NULL;
+    }
+
+    bool IsReadyToUpdate() const
+    {
+        spi_transaction_t t;
+        spi_transaction_t * tptr = &t;
+        esp_err_t ret = spi_device_get_trans_result(_spiHandle, &tptr, 0);
+
+        // We know the previous transaction completed if we got ESP_OK, and we know there's no transactions queued if tptr is unmodified
+        return (ret==ESP_OK || tptr == &t);
+    }
+
+    void Initialize(int8_t sck, int8_t dat0, int8_t dat1, int8_t dat2, int8_t dat3, int8_t ss)
+    {
+        memset(_data, 0x00, _sizeStartFrame);
+        memset(_data + _sizeStartFrame + _sizePixelData, 0x00, _spiBufferSize - (_sizeStartFrame + _sizePixelData));
+
+        _ssPin = ss;
+
+        esp_err_t ret;
+        spi_bus_config_t buscfg;
+        memset(&buscfg, 0x00, sizeof(buscfg));
+
+        buscfg.miso_io_num=dat1;
+        buscfg.mosi_io_num=dat0;
+        buscfg.sclk_io_num=sck;
+        buscfg.quadwp_io_num=dat2;
+        buscfg.quadhd_io_num=dat3;
+        buscfg.max_transfer_sz=_spiBufferSize;
+
+        //Initialize the SPI bus
+        ret=spi_bus_initialize(T_SPIBUS::spiHostDevice, &buscfg, T_SPIBUS::dmaChannel);
+        ESP_ERROR_CHECK(ret);
+
+        InitSpiDevice();
+    }
+
+    void Initialize(int8_t sck, int8_t miso, int8_t mosi, int8_t ss)
+    {
+        Initialize(sck, mosi, miso, -1, -1, ss);
+    }
+
+    // If pins aren't specified, initialize bus with just the default SCK and MOSI pins for the SPI peripheral (no SS, no >1-bit pins)
+    void Initialize()
+    {
+        if(T_SPIBUS::spiHostDevice == VSPI_HOST) {
+            Initialize(SCK, -1, MOSI, -1, -1, -1);
+        } else {
+            Initialize(14, -1, 13, -1, -1, -1);
+        }
+    }
+
+    void Update(bool)
+    {
+        while(!IsReadyToUpdate());
+
+        memcpy(_dmadata, _data, _spiBufferSize);
+
+        memset(&_spiTransaction, 0, sizeof(spi_transaction_t));
+        _spiTransaction.length=(_spiBufferSize) * 8; // in bits not bytes!
+        if(T_SPIBUS::parallelBits == 1)
```

Comment:
all branches must have space after them; "if(" should be "if (", makes it clear its not a function call.
again, branches must have {} even for simple cases like these.